In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [3]:
 #load the dataset 
df = pd.read_csv("Stock Prices Data Set.csv")
df.head()

symbol        date      open      high       low     close    volume
0    AAL  2014-01-02   25.0700   25.8200   25.0600   25.3600   8998943
1   AAPL  2014-01-02   79.3828   79.5756   78.8601   79.0185  58791957
2    AAP  2014-01-02  110.3600  111.8800  109.2900  109.7400    542711
3   ABBV  2014-01-02   52.1200   52.3300   51.5200   51.9800   4569061
4    ABC  2014-01-02   70.1100   70.2300   69.4800   69.8900   1148391

In [4]:
print(df.columns)

Index(['symbol', 'date', 'open', 'high', 'low', 'close', 'volume'], dtype='object')


In [5]:
df.describe()

open           high            low          close  \
count  497461.000000  497464.000000  497464.000000  497472.000000   
mean       86.352275      87.132562      85.552467      86.369082   
std       101.471228     102.312062     100.570957     101.472407   
min         1.620000       1.690000       1.500000       1.590000   
25%        41.690000      42.090000      41.280000      41.703750   
50%        64.970000      65.560000      64.353700      64.980000   
75%        98.410000      99.230000      97.580000      98.420000   
max      2044.000000    2067.990000    2035.110000    2049.000000   

             volume  
count  4.974720e+05  
mean   4.253611e+06  
std    8.232139e+06  
min    0.000000e+00  
25%    1.080166e+06  
50%    2.084896e+06  
75%    4.271928e+06  
max    6.182376e+08

In [6]:
print(df.shape)

(497472, 7)


In [7]:
print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 497472 entries, 0 to 497471
Data columns (total 7 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   symbol  497472 non-null  object 
 1   date    497472 non-null  object 
 2   open    497461 non-null  float64
 3   high    497464 non-null  float64
 4   low     497464 non-null  float64
 5   close   497472 non-null  float64
 6   volume  497472 non-null  int64  
dtypes: float64(4), int64(1), object(2)
memory usage: 26.6+ MB
None


In [8]:
#check for null values
df.isnull().sum()  #have zero null values

symbol     0
date       0
open      11
high       8
low        8
close      0
volume     0
dtype: int64

In [9]:
# Get the number of unique values in a specific column
unique_count = df['symbol'].nunique()
print(f"Number of unique values: {unique_count}")

Number of unique values: 505


Since i am provided with 500k rows and symbol have 505 unique values that means i am dealing with set of different companies and their traiding history all at once so i cannot sort the stocks based on date column as on same date i could have stocks of different companies or could mix say apple comancy stock price with google company .

Also the date column is of string variable which is needed to be transform into numeric value for further analysis.

In [10]:
# Convert text date to real date
df['date'] = pd.to_datetime(df['date'])

In [11]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 497472 entries, 0 to 497471
Data columns (total 7 columns):
 #   Column  Non-Null Count   Dtype         
---  ------  --------------   -----         
 0   symbol  497472 non-null  object        
 1   date    497472 non-null  datetime64[ns]
 2   open    497461 non-null  float64       
 3   high    497464 non-null  float64       
 4   low     497464 non-null  float64       
 5   close   497472 non-null  float64       
 6   volume  497472 non-null  int64         
dtypes: datetime64[ns](1), float64(4), int64(1), object(1)
memory usage: 26.6+ MB


In [12]:
top_50 = df['symbol'].value_counts().head(50)

print("Top 50 symbol and their Counts:")
print(top_50)

Top 50 symbol and their Counts:
symbol
YUM     1007
XYL     1007
XRX     1007
XRAY    1007
XOM     1007
XL      1007
XLNX    1007
XEL     1007
XEC     1007
WY      1007
WYN     1007
WYNN    1007
WU      1007
WM      1007
WMT     1007
WMB     1007
AMD     1007
AMAT    1007
ALXN    1007
ALL     1007
ALLE    1007
ALK     1007
ALGN    1007
ALB     1007
AKAM    1007
AJG     1007
AIZ     1007
AIV     1007
AIG     1007
AGN     1007
AFL     1007
AET     1007
ARE     1007
APH     1007
APD     1007
APC     1007
APA     1007
AOS     1007
AON     1007
ANTM    1007
ANSS    1007
ANDV    1007
AMZN    1007
AMT     1007
AMP     1007
AMG     1007
AMGN    1007
AME     1007
BDX     1007
BBY     1007
Name: count, dtype: int64


In [13]:
bottom_50 = df['symbol'].value_counts().tail(50)
print(bottom_50)

symbol
VRSK    1007
VNO     1007
VMC     1007
VLO     1007
VIAB    1007
ZTS     1007
ZION    1007
ZBH     1007
AES     1007
AEP     1007
AEE     1007
ADS     1007
ADSK    1007
ADP     1007
ADM     1007
ADI     1007
ADBE    1007
ACN     1007
ABT     1007
ABC     1007
ABBV    1007
AAP     1007
AAPL    1007
AAL     1007
ICE     1006
ES       997
DHR      995
O        995
GOOG     949
NAVI     934
INFO     891
SYF      862
CFG      824
QRVO     755
WRK      636
KHC      629
PYPL     629
HPE      555
HPQ      555
CSRA     535
WLTW     502
UA       438
FTV      378
EVHC     271
HLT      250
DXC      189
BHGE     126
BHF      117
DWDP      83
APTV      18
Name: count, dtype: int64


In [14]:
# Step 1: Count occurrences of each unique value
counts = df['symbol'].value_counts()

# Step 2: Count how many unique values have the same frequency
frequency_of_counts = counts.value_counts()

print("Frequency | Number of Unique Items with that Frequency")
print(frequency_of_counts)

Frequency | Number of Unique Items with that Frequency
count
1007    479
995       2
555       2
629       2
949       1
934       1
1006      1
997       1
862       1
891       1
755       1
824       1
636       1
535       1
502       1
438       1
378       1
271       1
250       1
189       1
126       1
117       1
83        1
18        1
Name: count, dtype: int64


In [15]:
# 1. Convert the 'date' column to a real datetime format
# (This fixes the "Object" data type issue)
df['date'] = pd.to_datetime(df['date'])

# 2. Filter the data to keep only Apple (AAPL)
# (We use .copy() to tell Python this is a new, separate table, not just a view of the old one)
df_clean = df[df['symbol'] == 'AAPL'].copy()

# 3. Sort chronologically from Oldest to Newest
# (Crucial for time-series. If this is wrong, 'Forward Fill' will fill from the future!)
df_clean = df_clean.sort_values('date')

# 4. Set the Date as the Index
# (Now the rows will be labeled by time, not 0, 1, 2...)
df_clean.set_index('date', inplace=True)

# 5. Check the result
print("--- Data Shape (Rows, Columns) ---")
print(df_clean.shape)
print("\n--- First 5 Rows ---")
df_clean.head()

--- Data Shape (Rows, Columns) ---
(1007, 6)

--- First 5 Rows ---


symbol     open     high      low    close     volume
date                                                            
2014-01-02   AAPL  79.3828  79.5756  78.8601  79.0185   58791957
2014-01-03   AAPL  78.9799  79.0999  77.2042  77.2828   98303870
2014-01-06   AAPL  76.7785  78.1142  76.2285  77.7042  103359151
2014-01-07   AAPL  77.7599  77.9942  76.8464  77.1481   79432766
2014-01-08   AAPL  76.9728  77.9371  76.9556  77.6371   64686685

In [18]:
# 1. Fill missing values using Forward Fill ('ffill')
# This propagates the last valid observation forward to next valid
df_clean = df_clean.fillna(method='ffill')

# 2. Double check: Are there any missing values left?
# (This should print all Zeros now)
print("--- Missing Values After Cleaning ---")
print(df_clean.isnull().sum())

--- Missing Values After Cleaning ---
symbol    0
open      0
high      0
low       0
close     0
volume    0
dtype: int64


C:\Users\shubh\AppData\Local\Temp\ipykernel_824\2003867035.py:3: FutureWarning: DataFrame.fillna with 'method' is deprecated and will raise in a future version. Use obj.ffill() or obj.bfill() instead.
  df_clean = df_clean.fillna(method='ffill')


In [19]:
from sklearn.preprocessing import MinMaxScaler

# 1. Initialize the Scaler (scales data between 0 and 1)
scaler = MinMaxScaler()

# 2. Apply the scaler to our clean data
# (Note: This returns a numpy array, so we convert it back to a DataFrame to keep column names)
df_scaled = pd.DataFrame(scaler.fit_transform(df_clean), 
                         columns=df_clean.columns, 
                         index=df_clean.index)

# 3. Split into Train (First 80%) and Test (Last 20%)
# We use .iloc to slice strictly by order, ensuring no shuffling happens
train_size = int(len(df_scaled) * 0.8)

train_data = df_scaled.iloc[:train_size]
test_data = df_scaled.iloc[train_size:]

# 4. Verify the split
print(f"Total Days: {len(df_scaled)}")
print(f"Training Days (First 80%): {len(train_data)}")
print(f"Testing Days (Last 20%):   {len(test_data)}")

# Show the scaled data (values should be between 0.0 and 1.0)
df_scaled.head()

ValueError: could not convert string to float: 'AAPL'

In [20]:
# 1. Drop the 'symbol' column (It's text, and we don't need it anymore)
# (We use errors='ignore' so it won't crash if you already dropped it)
df_clean = df_clean.drop('symbol', axis=1, errors='ignore')

# 2. Check to make sure only numbers are left
print("Columns left:", df_clean.columns.tolist())

# 3. NOW run the Scaling code again
from sklearn.preprocessing import MinMaxScaler

scaler = MinMaxScaler()
df_scaled = pd.DataFrame(scaler.fit_transform(df_clean), 
                         columns=df_clean.columns, 
                         index=df_clean.index)

# 4. Split into Train/Test again
train_size = int(len(df_scaled) * 0.8)
train_data = df_scaled.iloc[:train_size]
test_data = df_scaled.iloc[train_size:]

print("Success! Data Scaled and Split.")
df_scaled.head()

Columns left: ['open', 'high', 'low', 'close', 'volume']
Success! Data Scaled and Split.


open      high       low     close    volume
date                                                        
2014-01-02  0.082810  0.075114  0.080046  0.072566  0.185293
2014-01-03  0.078950  0.070607  0.064177  0.056039  0.340025
2014-01-06  0.057858  0.061269  0.054827  0.060052  0.359822
2014-01-07  0.067261  0.060132  0.060749  0.054757  0.266124
2014-01-08  0.059719  0.059591  0.061795  0.059413  0.208377

In [21]:
# Save the processed data to a new CSV file
# We include 'index=True' to keep the Dates!
df_scaled.to_csv('processed_aapl_data.csv')

print("Task 2 Completed! File saved as 'processed_aapl_data.csv'")

Task 2 Completed! File saved as 'processed_aapl_data.csv'
